apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,167 @@
+/*-
+ * #%L
+ * Elastic APM Java agent
+ * %%
+ * Copyright (C) 2018 - 2021 Elastic and contributors
+ * %%
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * #L%
+ */
+package co.elastic.apm.agent.javalin;
+
+import co.elastic.apm.agent.bci.TracerAwareInstrumentation;
+import co.elastic.apm.agent.impl.transaction.AbstractSpan;
+import co.elastic.apm.agent.impl.transaction.Span;
+import co.elastic.apm.agent.impl.transaction.Transaction;
+import co.elastic.apm.agent.util.VersionUtils;
+import io.javalin.http.Context;
+import io.javalin.http.Handler;
+import io.javalin.http.HandlerType;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+
+import javax.annotation.Nullable;
+import java.lang.reflect.Field;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.concurrent.CompletableFuture;
+
+import static co.elastic.apm.agent.impl.transaction.AbstractSpan.PRIO_HIGH_LEVEL_FRAMEWORK;
+import static net.bytebuddy.matcher.ElementMatchers.hasSuperType;
+import static net.bytebuddy.matcher.ElementMatchers.isInterface;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.not;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+
+public class JavalinInstrumentation extends TracerAwareInstrumentation {
+
+    private static final String FRAMEWORK_NAME = "Javalin";
+
+    @Override
+    public ElementMatcher<? super TypeDescription> getTypeMatcher() {
+        return hasSuperType(named("io.javalin.http.Handler")).and(not(isInterface()));
```

Comment:
Hm. I don't have any good idea to limit this TBH, as this is just a functional interface and I have a hard time assuming that everyone suffixes/prefixes their javalin class with `Handler` - I saw in [JaxRsTransactionNameInstrumentation](https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugins/apm-jaxrs-plugin/src/main/java/co/elastic/apm/agent/jaxrs/JaxRsTransactionNameInstrumentation.java#L70-L75) that might help if application packages are specified. Should I go that route?

Reply:
Overriding `co.elastic.apm.agent.sdk.ElasticApmInstrumentation#getClassLoaderMatcher` to restrict only to class loaders that can load the `io.javalin.http.Handler` interface is already a great improvement. It relies on the fact that implementations must be loaded from a class loader to which the interface is visible.

The `application_packages` is a good option if all `Handler` implementations are expected to be such. However, it doesn't fit very well with our aim to get a good zero-configuration experience.

Maybe the best option is to use the [`ensureInstrumented`](https://github.com/elastic/apm-agent-java/blob/master/apm-agent-plugin-sdk/src/main/java/co/elastic/apm/agent/sdk/DynamicTransformer.java#L47) capability, which dynamically instruments classes if they are not yet instrumented, but it requires that you instrument the place where the handler is submitted for execution. Your `addHandler` instrumentation would only cover agent attachment at startup, and not runtime attachment. If you find some `submit(Handler handler)` or `invoke(Handler handler)` API to instrument, that would be good.

Reply:
I just looked at the code, it looks like `addHandler` is not some type of registration, it is called for every handling, is that correct? If so, it is a good candidate to do `ensureInstrumented`

Reply:
You may be confusing the two intrumentations here:

1. `JavalinHandlerLambdaInstrumentation` is instrumented against `io.javalin.Javalin.addHandler()` and is only called when a new handler is added in order to wrap anonymous lambdas
2. `JavalinInstrumentation` instruments `io.javalin.http.Context.handle()` and is invoked on every HTTP request, so this would indeed be a good candidate for ensureInstrumented instrumentation? However how much sense does this make, if the lambda wrapping is not applied, so it would only apply for non-lambda anonymous handlers in that case?

Reply:
No, I am not confusing, but I think it is the time for a chat, rather than this back and forth 🙂

Reply:
Chat summary - we'll start with the class loader matcher only and that should be OK